src/polymer/polymer2-element-scanner.ts

```diff
@@ -436,66 +318,139 @@ class ElementVisitor implements Visitor {
-        if (tagNameResult.successful) {
-          element.tagName = tagNameResult.value;
-        } else {
-          this._warnings.push(tagNameResult.value);
-        }
-        this._elements.add(element);
+  ast: estree.Node;
+}
+class ClassFinder implements Visitor {
+  readonly classes: FoundClass[] = [];
+  private readonly alreadyMatched = new Set<estree.ClassExpression>();
+
+  enterAssignmentExpression(
+      node: estree.AssignmentExpression, parent: estree.Node) {
+    this.handleGeneralAssignment(
+        astValue.getIdentifierName(node.left), node.right, node, parent);
+  }
+
+  enterVariableDeclarator(
+      node: estree.VariableDeclarator, parent: estree.Node) {
+    if (node.init) {
+      this.handleGeneralAssignment(
+          astValue.getIdentifierName(node.id), node.init, node, parent);
+    }
+  }
+
+  private handleGeneralAssignment(
+      assignedName: string|undefined, value: estree.Expression,
+      assignment: estree.VariableDeclarator|estree.AssignmentExpression,
+      statement: estree.Node) {
+    const comment = esutil.getAttachedComment(value) ||
+        esutil.getAttachedComment(assignment) ||
+        esutil.getAttachedComment(statement) || '';
+    const doc = jsdoc.parseJsdoc(comment);
+    if (value.type === 'ClassExpression') {
+      const name =
+          assignedName || value.id && astValue.getIdentifierName(value.id);
+
+      this._classFound(name, doc, value);
+    } else {
+      // TODO(rictic): remove @polymerElement here.
+      //     See: https://github.com/Polymer/polymer-analyzer/issues/605
+      if (jsdoc.hasTag(doc, 'customElement') ||
+          jsdoc.hasTag(doc, 'polymerElement')) {
+        this._classFound(assignedName, doc, value);
       }
     }
-    return Array.from(this._elements);
   }
 
-  getTagNameFromExpression(expression: TagNameExpression):
-      Result<string|undefined, Warning> {
-    if (expression.type === 'string-literal') {
-      return {successful: true, value: expression.value};
+  enterClassExpression(node: estree.ClassExpression, parent: estree.Node) {
+    if (this.alreadyMatched.has(node)) {
+      return;
+    }
+
+    const name = node.id && astValue.getIdentifierName(node.id);
+    const comment = esutil.getAttachedComment(node) ||
+        esutil.getAttachedComment(parent) || '';
+    this._classFound(name, jsdoc.parseJsdoc(comment), node);
+  }
+
+  enterClassDeclaration(node: estree.ClassDeclaration, parent: estree.Node) {
+    const name = astValue.getIdentifierName(node.id);
+    const comment = esutil.getAttachedComment(node) ||
+        esutil.getAttachedComment(parent) || '';
+    this._classFound(name, jsdoc.parseJsdoc(comment), node);
+  }
+
+  private _classFound(
+      name: string|undefined, doc: jsdoc.Annotation, ast: estree.Node) {
+    const namespacedName = name && getNamespacedIdentifier(name, doc);
+
+    this.classes.push({name, namespacedName, doc, ast});
+    if (ast.type === 'ClassExpression') {
```

Comment:
can you explain this, and why `enterClassDeclaration` doesn't check for alreadyMatched`, in documentation?